### PR TITLE
output: fix io.Reader when used with LineMap

### PR DESCRIPTION
**Pull request by bobheadxi**
I noticed LineMaps are not applied in my "fix" for io.Reader in which I kind of completely forgot about LineMaps.

1. Fix io.Reader + line maps
1. More robust tests for various line map scenarios

The implementation is not ideal, there are quite a few layers of buffering. I explored incremental reading of the command output, but it's super hard (impossible?) without lookahead, since we never know when the line has really ended, and LineMap depends on full lines

I wonder if we can lift line map assignment out of output, and into command. Then we can perhaps deal purely with mapped output 

tldr I might have overcomplicated things 😂😂😂

### CODE REVIEW

File: command_test.go [hunk 1]
```diff
@@ -38,57 +36,115 @@ func TestRunAndAggregate(t *testing.T) {
...
+			output: func() run.Output {
+				return run.Cmd(ctx, command).Run().
+					Map(func(ctx context.Context, line []byte, dst io.Writer) (int, error) {
+						return dst.Write(bytes.ReplaceAll(line, []byte("hello"), []byte("goodbye")))
+					}).
+					Map(func(ctx context.Context, line []byte, dst io.Writer) (int, error) {
+						return dst.Write(bytes.ReplaceAll(line, []byte("world"), []byte("jh")))
+					})
+			},
+			expect: "goodbye jh",
```
Comment (jhchabran) [inline command_test.go:134]: Technically, it's good morning here 😆 🤗

File: buffer.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+)
+
+// maxBufferSize denotes the maximum size of each buffer. Overflows are written to disk
+// at increments of this size.
+var maxBufferSize int64 = 128 * 1024
+
+// bufferPool will never return an error.
+//
+// Uses unbounded buffers that create files of size fileBuffersSize to store overflow.
+func makeBuffer() buffer.Buffer {
```
Comment (jhchabran) [inline buffer.go:14]:
```suggestion
func makeUnboundedBuffer() buffer.Buffer {
```

Could we rename this func to ensure we don't get puzzled when reading the code? With all the other buffer uses, it's not obvious why we would use ours here by just reading the function name.

**Review by jhchabran — APPROVED** [review]
Indeed this is getting complicated. But that's fine as we're learning as we go through. Trying to see how we can refactor all of it during the offsite would be a cool activity 😊
